File: python/trello.py
#!/usr/bin/env python3

import os
import sys
import json
import yaml
import requests
from pathlib import Path
from dotenv import load_dotenv

class TrelloAPI:
    BASE_URL = 'https://api.trello.com/1'
    CONFIG_FILE = Path.home() / '.trello_config.yml'
    
    def __init__(self):
        load_dotenv()
        self.api_key = os.getenv('TRELLO_API_KEY')
        self.secret = os.getenv('TRELLO_SECRET')
        
        if not self.api_key or not self.secret:
            # Try to read directly from .env file
            if os.path.exists('.env'):
                with open('.env', 'r') as f:
                    for line in f:
                        if line.startswith('TRELLO_API_KEY='):
                            self.api_key = line.strip().split('=')[1]
                        elif line.startswith('TRELLO_SECRET='):
                            self.secret = line.strip().split('=')[1]
        
        if not self.api_key:
            raise ValueError("TRELLO_API_KEY not found in .env file or environment variables")
        if not self.secret:
            raise ValueError("TRELLO_SECRET not found in .env file or environment variables")
        
        self.load_or_create_config()

    def load_or_create_config(self):
        if self.CONFIG_FILE.exists():
            with open(self.CONFIG_FILE, 'r') as f:
                config = yaml.safe_load(f)
                self.token = config['token']
                self.board_id = config['board_id']
                self.board_name = config['board_name']
        else:
            # Get token through authorization flow
            auth_url = f"https://trello.com/1/authorize?expiration=never&scope=read,write&response_type=token&name=Server%20Token&key={self.api_key}"
            print("\nPlease visit this URL to authorize the application:")
            print(auth_url)
            self.token = input("\nEnter the token you received: ").strip()

            # Get boards and select one
            boards = self.get_boards()
            self.print_boards(boards)
            board_index = int(input("\nSelect board number: ")) - 1
            board = boards[board_index]
            self.board_id = board['id']
            self.board_name = board['name']

            # Save config
            self.save_config()

    def save_config(self):
        config = {
            'token': self.token,
            'board_id': self.board_id,
            'board_name': self.board_name
        }
        self.CONFIG_FILE.parent.mkdir(parents=True, exist_ok=True)
        with open(self.CONFIG_FILE, 'w') as f:
            yaml.dump(config, f)

    def get_boards(self):
        response = requests.get(
            f"{self.BASE_URL}/members/me/boards",
            params={
                'key': self.api_key,
                'token': self.token,
                'fields': 'name,id'
            }
        )
        return response.json()

    def get_lists(self):
        response = requests.get(
            f"{self.BASE_URL}/boards/{self.board_id}/lists",
            params={
                'key': self.api_key,
                'token': self.token,
                'fields': 'name,id'
            }
        )
        return response.json()

    def get_all_cards(self):
        response = requests.get(
            f"{self.BASE_URL}/boards/{self.board_id}/cards",
            params={
                'key': self.api_key,
                'token': self.token,
                'fields': 'name,id,idList'
            }
        )
        return response.json()

    def move_card(self, card_id, list_id):
        response = requests.put(
            f"{self.BASE_URL}/cards/{card_id}",
            params={
                'key': self.api_key,
                'token': self.token,
                'idList': list_id
            }
        )
        return response.json()

    def print_boards(self, boards):
        print("\nAvailable boards:")
        for i, board in enumerate(boards, 1):
            print(f"{i}. {board['name']}")

def fuzzy_match(needle, haystack):
    needle = needle.lower()
    matches = [item for item in haystack if needle in item['name'].lower()]
    
    if not matches:
        print(f"\nNo matches found for '{needle}'")
        print("Available options:")
        for item in haystack:
            print(f"- {item['name']}")
        sys.exit(1)
    elif len(matches) == 1:
        return matches[0]
    else:
        print(f"\nMultiple matches found for '{needle}':")
        for i, match in enumerate(matches, 1):
            print(f"{i}. {match['name']}")
        index = int(input("Select number: ")) - 1
        return matches[index]

def main():
    if len(sys.argv) != 3:
        print(f"Usage: {sys.argv[0]} <card_name> <column>")
        print(f"e.g. {sys.argv[0]} fcas testing")
        sys.exit(1)

    card_name = sys.argv[1]
    target_column = sys.argv[2]
    
    trello = TrelloAPI()
    
    # Get all lists and cards
    lists = trello.get_lists()
    cards = trello.get_all_cards()
    
    # Find card first
    card = fuzzy_match(card_name, cards)
    
    # Find target list
    target_list = fuzzy_match(target_column, lists)
    
    # Move the card
    if card['idList'] == target_list['id']:
        print(f"\nCard '{card['name']}' is already in list '{target_list['name']}'")
    else:
        current_list = next(l for l in lists if l['id'] == card['idList'])
        trello.move_card(card['id'], target_list['id'])
        print(f"\nMoved card '{card['name']}' from '{current_list['name']}' to '{target_list['name']}'")

if __name__ == "__main__":
    main() 

File: ruby/trello.rb
#!/usr/bin/env ruby

require 'dotenv/load'
require 'httparty'
require 'json'
require 'yaml'
require 'fileutils'

class TrelloAPI
  BASE_URL = 'https://api.trello.com/1'
  CONFIG_FILE = File.expand_path('~/.trello_config.yml')
  
  def initialize
    @api_key = ENV['TRELLO_API_KEY']
    @secret = ENV['TRELLO_SECRET']
    
    if @api_key.nil? || @secret.nil?
      # Try to read directly from .env file
      if File.exist?('.env')
        File.readlines('.env').each do |line|
          if line.start_with?('TRELLO_API_KEY=')
            @api_key = line.strip.split('=')[1]
          elsif line.start_with?('TRELLO_SECRET=')
            @secret = line.strip.split('=')[1]
          end
        end
      end
    end
    
    raise "TRELLO_API_KEY not found in .env file or environment variables" unless @api_key
    raise "TRELLO_SECRET not found in .env file or environment variables" unless @secret
    
    load_or_create_config
  end

  def load_or_create_config
    if File.exist?(CONFIG_FILE)
      config = YAML.load_file(CONFIG_FILE)
      @token = config['token']
      @board_id = config['board_id']
      @board_name = config['board_name']
    else
      # Get token through authorization flow
      auth_url = "https://trello.com/1/authorize?expiration=never&scope=read,write&response_type=token&name=Server%20Token&key=#{@api_key}"
      puts "\nPlease visit this URL to authorize the application:"
      puts auth_url
      print "\nEnter the token you received: "
      @token = STDIN.gets.strip

      # Get boards and select one
      boards = get_boards
      print_boards(boards)
      print "\nSelect board number: "
      board_index = STDIN.gets.to_i - 1
      board = boards[board_index]
      @board_id = board['id']
      @board_name = board['name']

      # Save config
      save_config
    end
  end

  def save_config
    config = {
      'token' => @token,
      'board_id' => @board_id,
      'board_name' => @board_name
    }
    FileUtils.mkdir_p(File.dirname(CONFIG_FILE))
    File.write(CONFIG_FILE, config.to_yaml)
  end

  def get_boards
    response = HTTParty.get("#{BASE_URL}/members/me/boards", query: {
      key: @api_key,
      token: @token,
      fields: 'name,id'
    })
    JSON.parse(response.body)
  end

  def get_lists
    response = HTTParty.get("#{BASE_URL}/boards/#{@board_id}/lists", query: {
      key: @api_key,
      token: @token,
      fields: 'name,id'
    })
    JSON.parse(response.body)
  end

  def get_all_cards
    response = HTTParty.get("#{BASE_URL}/boards/#{@board_id}/cards", query: {
      key: @api_key,
      token: @token,
      fields: 'name,id,idList'
    })
    JSON.parse(response.body)
  end

  def move_card(card_id, list_id)
    response = HTTParty.put("#{BASE_URL}/cards/#{card_id}", query: {
      key: @api_key,
      token: @token,
      idList: list_id
    })
    JSON.parse(response.body)
  end

  def print_boards(boards)
    puts "\nAvailable boards:"
    boards.each_with_index do |board, index|
      puts "#{index + 1}. #{board['name']}"
    end
  end
end

def fuzzy_match(needle, haystack)
  needle = needle.downcase
  matches = haystack.select do |item|
    item['name'].downcase.include?(needle)
  end
  
  case matches.size
  when 0
    puts "\nNo matches found for '#{needle}'"
    puts "Available options:"
    haystack.each { |item| puts "- #{item['name']}" }
    exit 1
  when 1
    matches.first
  else
    puts "\nMultiple matches found for '#{needle}':"
    matches.each_with_index do |match, i|
      puts "#{i + 1}. #{match['name']}"
    end
    print "Select number: "
    index = STDIN.gets.to_i - 1
    matches[index]
  end
end

def main
  if ARGV.length != 2
    puts "Usage: #{$0} <card_name> <column>"
    puts "e.g. #{$0} fcas testing"
    exit 1
  end

  card_name = ARGV[0]
  target_column = ARGV[1]
  
  trello = TrelloAPI.new
  
  # Get all lists and cards
  lists = trello.get_lists
  cards = trello.get_all_cards
  
  # Find card first
  card = fuzzy_match(card_name, cards)
  
  # Find target list
  target_list = fuzzy_match(target_column, lists)
  
  # Move the card
  if card['idList'] == target_list['id']
    puts "\nCard '#{card['name']}' is already in list '#{target_list['name']}'"
  else
    current_list = lists.find { |l| l['id'] == card['idList'] }
    trello.move_card(card['id'], target_list['id'])
    puts "\nMoved card '#{card['name']}' from '#{current_list['name']}' to '#{target_list['name']}'"
  end
end

if __FILE__ == $0
  main
end